src/java.base/share/native/launcher/main.c

```diff
@@ -278,15 +321,32 @@ main(int argc, char **argv)
         margv = args->elements;
     }
 
-    // Avoid unexpected process completion when checkpointing under docker container run
-    // by creating the main process waiting for children before exit.
-    if (is_checkpoint && 1 == getpid()) {
-        g_child_pid = fork();
-        if (0 < g_child_pid) {
-            // The main process should forward signals to the child.
-            setup_sighandler();
-            const int status = wait_for_children();
-            exit(status);
+    if (is_checkpoint) {
+        const int crac_min_pid_default = 128;
+        const char *env_min_pid_str = getenv("CRAC_MIN_PID");
+        const int env_min_pid = env_min_pid_str ? atoi(env_min_pid_str) : 0;
+        // TODO: should it be checked for max pid overflow?
+        const int crac_min_pid = 0 < env_min_pid ? env_min_pid : crac_min_pid_default;
+
+        if (getpid() <= crac_min_pid) {
+            // Move PID value for new processes to a desired value
+            // to avoid PID conflicts on restore.
+            if (get_last_pid() < crac_min_pid) {
+                set_last_pid(crac_min_pid);
+                if (get_last_pid() < crac_min_pid) {
+                    spin_last_pid(crac_min_pid);
+                }
+            }
+
+            // Avoid unexpected process completion when checkpointing under docker container run
+            // by creating the main process waiting for children before exit.
+            g_child_pid = fork();
+            if (0 < g_child_pid) {
```

Comment:
Does it make sense to check if `g_child_pid < crac_min_pid`? As the ns_last_pid is a subject for races, and instead of guessing what the next pid will be, we'll have the ability to explicity check if the child has the right pid. The child will have to compare its pid with the expected crac_min_pid.

Reply:
I'm not sure I understand your thoughts. Could you explain what do you mean, please?

`if (0 < g_child_pid)` here is just a detector for child/parent code to execute, nothing more. At this line, there is no need for child to check its PID.

Reply:
I mean to check child pid (that is `g_check_pid), does it satisfy our min pid requirements. 

Right now, we have to check last_ns_pid and _assume_ what will be the child pid. Instead, we may create the child and based on the pid it gets, decide if the requirement is satifised, or we need another fork. Althought this will require one extra fork, if last_ns_pid is writeable.

This actually relates to https://github.com/openjdk/crac/pull/86#discussion_r1236956565.
OK, please disregard this.